sdk/scheduler/src/main/java/com/mesosphere/sdk/storage/PersisterCache.java

```diff
@@ -0,0 +1,121 @@
+package com.mesosphere.sdk.storage;
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A transparent write-through cache for an underlying {@link Persister} instance. Each cache instance is thread-safe,
+ * but there is no guarantee of consistent behavior across multiple cache instances.
+ */
+public class PersisterCache implements Persister {
+
+    private static final Logger logger = LoggerFactory.getLogger(PersisterCache.class);
+
+    private final ReadWriteLock internalLock = new ReentrantReadWriteLock();
+    private final Lock rlock = internalLock.readLock();
+    private final Lock rwlock = internalLock.writeLock();
+
+    private final Persister persister;
+    private MemPersister cache;
+
+    public PersisterCache(Persister persister) throws PersisterException {
+        this.persister = persister;
+        refresh();
+    }
+
+    @Override
+    public byte[] get(String path) throws PersisterException {
+        rlock.lock();
+        try {
+            return cache.get(path);
+        } finally {
+            rlock.unlock();
+        }
+    }
+
+    @Override
+    public Collection<String> getChildren(String path) throws PersisterException {
+        rlock.lock();
+        try {
+            return cache.getChildren(path);
+        } finally {
+            rlock.unlock();
+        }
+    }
+
+    @Override
+    public void set(String path, byte[] bytes) throws PersisterException {
+        rwlock.lock();
+        try {
+            persister.set(path, bytes);
```

Comment:
The cache.set() call is unlikely to fail as it's seeking/setting a value in a local tree structure. Failure there would imply OOM